exporter/splunkhecexporter/client.go

```diff
@@ -260,20 +271,24 @@ func (c *client) fillMetricsBuffer(metrics pmetric.Metrics, buf buffer, is iterS
 
 				// Parsing metric record to Splunk event.
 				events := mapMetricToSplunkEvent(rm.Resource(), metric, c.config, c.logger)
-				tempBuf.Reset()
+
 				for _, event := range events {
 					// JSON encoding event and writing to buffer.
-					b, err := marshalEvent(event, c.config.MaxEventSize, jsonStream)
-					if err != nil {
-						permanentErrors = append(permanentErrors, consumererror.NewPermanent(fmt.Errorf("dropped metric event: %v, error: %w", event, err)))
+					jsonStream.Error = nil
+					jsonStream.WriteVal(event)
+					if jsonStream.Error != nil {
+						permanentErrors = append(permanentErrors, consumererror.NewPermanent(fmt.Errorf("dropped metric event: %v, error: %w", event, jsonStream.Error)))
 						continue
 					}
-					tempBuf.Write(b)
+					if uint(jsonStream.Buffered()) > c.config.MaxEventSize {
+						permanentErrors = append(permanentErrors, fmt.Errorf("event size %d exceeds limit %d", jsonStream.Buffered(), c.config.MaxEventSize))
+						continue
+					}
```

Comment:
I don't think this approach works for metrics. We are not flushing or resetting the stream once this condition is triggered. This means we will be dropping all the events after we have buffered more than `c.config.MaxEventSize`. This is incorrect because `MaxEventSize` is supposed to be applied to each individual event.

One way forward I can think of is introducing another field to `iterState` for "datapoints" and using it in this loop. Having that, we don't need to remove `marshalEvent` (or whatever we call it) and copy-paste the code. Later we can use it in `mapMetricToSplunkEvent` as well to avoid redoing translations.

Reply:
ok let me drop that code path for now, and we can revisit when we have exporterhelper and a way to keep track of event consumption.

Reply:
OK changes made - the optimization is much less important, we get 3% only, but hopefully it helps for other cases.

Reply:
Looks like the number of allocations is the same. In some cases, we utilize less memory overall and cpu, but it doesn't seem too much. At the same time, we make the code less readable. I'm not 100% sure if this change is a good improvement, considering a hypothetical error budget